src/services/FirestoreService.ts

```diff
@@ -0,0 +1,309 @@
+   *
+   * @returns Observable<IFirebaseDocument[]>
+   */
+  public subscribeToAllDealsForUser(): Observable<Array<IFirebaseDocument>> {
+    const allPublicDeals = this.getQueryObservable<Array<IFirebaseDocument>>(allPublicDealsQuery);
+
+    const queries$$: Subject<Observable<Array<IFirebaseDocument>>> = new Subject();
+
+    this.firebaseService.authStateChanged().subscribe(
+      user => {
+        if (!user) {
+          queries$$.next(allPublicDeals);
+        } else {
+          const representativeDeals = this.getQueryObservable<Array<IFirebaseDocument>>(representativeDealsQuery(user.uid));
+          const proposalLeadDeals = this.getQueryObservable<Array<IFirebaseDocument>>(proposalLeadDealsQuery(user.uid));
+          queries$$.next(combineLatest([
+            allPublicDeals,
+            representativeDeals,
+            proposalLeadDeals,
+          ]).pipe(
+            map(data => Utils.uniqById<IFirebaseDocument>(data.flat())),
+          ));
+        }
+      },
+    );
+
+    const queries$: Observable<Array<IFirebaseDocument>> = queries$$.pipe(
+      mergeAll(),
+    );
+
+    return queries$;
+  }
+
+  /**
+   * Observes all public deals in Firestore
+   *
+   * @returns Observable<IFirebaseDocument[]>
+   */
+  public subscribeToAllPublicDeals(): Observable<Array<IFirebaseDocument>> {
+    return this.getQueryObservable<Array<IFirebaseDocument>>(allPublicDealsQuery);
+  }
+
+  /**
+   * Adds deal clause -> discussion to Firestore discussions map
+   *
+   * @param dealId string
+   * @param clauseId string
+   * @param discussionHash string
+   * @returns Promise<void>
+   */
+  public async addClauseDiscussion(dealId: string, clauseId: string, discussionHash: string) {
+    try {
+      const ref = doc(firebaseDatabase, DEALS_COLLECTION, dealId);
+      await setDoc(
+        ref,
+        {
+          discussions: {[clauseId]: discussionHash},
+        },
+        {merge: true},
+      );
+    } catch (error) {
+      throw new Error(error);
+    }
+  }
+
+  /**
+   * Gets a RxJS Observable for Firestore Query.
+   *
+   * Turns Firestore onSnapshot method into Observable using RxJS fromEventPattern
+   *
+   * @param q Query<DocumentData>
+   * @returns Observable
+   */
+  private getQueryObservable<T>(q: Query<DocumentData>): Observable<T> {
+    return fromEventPattern(
+      (handler) => onSnapshot(
+        q,
+        (querySnapshot: QuerySnapshot<DocumentData>) => {
+          handler(this.getDocumentsFromQuerySnapshot(querySnapshot));
+        },
+      ),
+      (handler, unsubscribe: Unsubscribe) => {
+        unsubscribe();
+      },
+    );
+  }
+
+  /**
+   * Turns Firestore QuerySnapshot into a collection
+   *
+   * @param querySnapshot QuerySnapshot<DocumentData>
+   * @returns IFirebaseDocument[]
+   */
+  private getDocumentsFromQuerySnapshot(querySnapshot: QuerySnapshot<DocumentData>): Array<IFirebaseDocument> {
```

Comment:
maybe we can add this to the "Shared" directory as well (only if the "Shared" directory idea works in the first place 😄 )